src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxTransformNode.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text;
+using System.Threading;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal interface ISyntaxTransformNode
+    {
+        ISyntaxTransformBuilder GetBuilder(DriverStateTable previousStateTable);
+    }
+
+    internal interface ISyntaxTransformBuilder
+    {
+        bool Filter(SyntaxNode node);
+
+        void AddFilterFromPreviousTable(SemanticModel model, EntryState state);
+
+        void AddFilterEntries(ImmutableArray<SyntaxNode> nodes, SemanticModel model);
+
+        void SetInputState(DriverStateTable.Builder builder);
+    }
+
+    internal sealed class SyntaxTransformNode<T> : IIncrementalGeneratorNode<T>, ISyntaxTransformNode
+    {
+        private readonly Func<GeneratorSyntaxContext, T> _func;
+        private readonly IEqualityComparer<T> _comparer;
+        private readonly Func<SyntaxNode, bool> _filterFunc;
+
+        private readonly InputNode<SyntaxNode> _filterNode = new InputNode<SyntaxNode>();
+
+        internal SyntaxTransformNode(Func<SyntaxNode, bool> filterFunc, Func<GeneratorSyntaxContext, T> transformFunc, IEqualityComparer<T>? comparer = null)
+        {
+            _func = transformFunc;
+            _filterFunc = filterFunc;
+            _comparer = comparer ?? EqualityComparer<T>.Default;
+        }
+
+        // this is an input node, so we don't perform any updates
+        public NodeStateTable<T> UpdateStateTable(DriverStateTable.Builder graphState, NodeStateTable<T> previousTable, CancellationToken cancellationToken) => previousTable;
```

Comment:
Do we ever actually call this method for a `SyntaxTransformNode`? If so I'm not really understanding why we don't update the state table here. The other nodes that use filters, like `TransformNode` do their filtering in this method.

Reply:
Or perhaps should input nodes have a different abstraction than incremental nodes? Essentially should they have an interface which doesn't have an `UpdateStateTable` method?

Reply:
So the majority of nodes calculate their state recursively in response to a downstream node asking for it: the driverstatebuilder calls into `UpdateStateTable`  to get the updated state,  the node calculates its own state by getting its upstream, and returns it to the downstream one. These are updates 'in' the graph. The input nodes don't calculate their state in response to a downstream node because they're terminal: there is no 'upstream' for them to calculate from. Instead the graph sets their value 'outside' before running any updates. When the input node is eventually asked for its state by a downstream it can just return its 'previous' state (a misnomer, because its really the state that was set 'outside' before the graph ran)

---
In reply to: [627065607](https://github.com/dotnet/roslyn/pull/53147#discussion_r627065607) [](ancestors = 627065607)

Reply:
Yes, at this point I think they are really two different enough concepts that we should split them out.

---
In reply to: [627068217](https://github.com/dotnet/roslyn/pull/53147#discussion_r627068217) [](ancestors = 627068217)

Reply:
Will do this in a follow up PR

---
In reply to: [627656677](https://github.com/dotnet/roslyn/pull/53147#discussion_r627656677) [](ancestors = 627656677,627068217)

Reply:
This is addressed in a follow up PR

---
In reply to: [627656262](https://github.com/dotnet/roslyn/pull/53147#discussion_r627656262) [](ancestors = 627656262,627065607)